src/Cryptography/Sodium/Bindings/Signing.hs

```diff
@@ -0,0 +1,276 @@
+    -- $introduction
+
+    -- * Key pair generation
+    cryptoSignKeyPair,
+    cryptoSignSeedKeyPair,
+
+    -- * Combined mode
+    cryptoSign,
+    cryptoSignOpen,
+
+    -- * Detached Mode
+    cryptoSignDetached,
+
+    -- * Multi-part messages
+    -- $mpm
+    CryptoSignState,
+    cryptoSignStateBytes,
+    withSignState,
+    cryptoSignInit,
+    cryptoSignUpdate,
+    cryptoSignFinalCreate,
+    cryptoSignFinalVerify,
+    cryptoSignED25519SkToSeed,
+    cryptoSignED25519SkToPk,
+
+    -- * Constants
+    cryptoSignPublicKeyBytes,
+    cryptoSignSecretKeyBytes,
+    cryptoSignSecretBytes,
+    cryptoSignSeedBytes,
+  )
+where
+
+import Foreign (Ptr, allocaBytes)
+import Foreign.C (CInt (..), CSize (..), CUChar (..), CULLong (..))
+
+-- $introduction
+--
+-- When signing with public-key cryptography,  a signer generates a key pair consisting of:
+--   * A secret key, which you can use to append a signature to any number of messages.
+--   * A public key, which anybody can use to verify that the signature appended to a
+--     message was issued by the creator of the public key.
+--
+-- Verifiers need to already know and ultimately trust a public key before messages signed using
+-- it can be verified.
+-- Warning: this is different from authenticated encryption. Appending a signature does not change
+-- the representation of the message itself.
+
+-------------------------
+-- Key pair generation --
+-------------------------
+
+-- | This function randomly generates a secret key and a corresponding public key.
+-- The public key is put into the @pk@ parameter and the secret key into the @sk@ parameter.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_sign_keypair"
+  cryptoSignKeyPair ::
+    -- | @pk@ parameter. It has a length of 'cryptoSignPublicKeyBytes'.
+    Ptr CUChar ->
+    -- | @sk@ parameter. It has a length of 'cryptoSignSecretKeyBytes'.
+    Ptr CUChar ->
+    -- | Return code is 0 on success, -1 on error.
+    IO CInt
+
+-- | This function derives a keypair (@sk@ and @pk@) from a seed.
+-- It is deterministic.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_sign_seed_keypair"
+  cryptoSignSeedKeyPair ::
+    -- | @pk@ parameter. It has a length of 'cryptoSignPublicKeyBytes'.
+    Ptr CUChar ->
+    -- | @sk@ parameter. It has a length of 'cryptoSignSecretKeyBytes'.
+    Ptr CUChar ->
+    -- | @seed@ parameter. It has a length of 'cryptoSignSeedBytes'.
+    Ptr CUChar ->
+    -- | Return code is 0 on success, -1 on error.
+    IO CInt
+
+-------------------
+-- Combined Mode --
+-------------------
+
+-- | The 'cryptoSign' function prepends a signature to a message @m@, whose length is @mlen@ bytes,
+-- using the secret key @sk@.
+--
+-- The signed message, which includes the signature plus an unaltered copy of the message, is put
+-- into @sm@ and is 'cryptoSignBytes' + @mlen@ bytes long.
+--
+-- If @smlen@ is not a @NULL@ pointer, then the actual length of the signed message is stored in
```

Comment:
I think 'If `smlen` is not `nullPtr`' makes far more sense in a Haskell context.